Rewrite openqa-advanced-retrigger-jobs in python

okurz opened the pull request.

okurz committed:
openqa-advanced-retrigger-jobs: Add usage+help

okurz committed:
Rewrite openqa-advanced-retrigger-jobs in python

perlpunk reviewed (commented):
It would still be good to have an example. I remember that it was complicated to get the quoting right

okurz commented:
> It would still be good to have an example. I remember that it was complicated to get the quoting right

An example for what? Do you mean an advanced example call in the help text?

asdil12 commented on openqa-advanced-retrigger-jobs:
```diff
@@ -1,16 +1,88 @@
+    )
+
+    log.debug(f"Using SQL query: '{query}' on {args.host}")
+    ssh_command = f'ssh {args.host} "sudo -u geekotest psql --no-align --tuples-only --command=\\"{query}\\" openqa"'
```
Better use an array instead of a space separated string, otherwise a shell needs to parse it.

okurz replied:
done

perlpunk commented:
> An example for what? Do you mean an advanced example call in the help text?

yes

perlpunk commented:
Also how about moving scripts to `bin/` and adding an extension, `.py`. I don't have a problem with extensions, and it makes tooling like linting easier

okurz commented:
> Also how about moving scripts to `bin/` and adding an extension, `.py`. I don't have a problem with extensions, and it makes tooling like linting easier

An extension would mean that we would need a separate installation tooling and step to install as executable without extension and I would like to avoid that.

okurz commented:
> It would still be good to have an example. I remember that it was complicated to get the quoting right

Done

sqozz commented on openqa-investigate:
```diff
@@ -1,391 +1,178 @@
-}
-
-caller 0 > /dev/null || main "$@"
+import requests
```
You're missing a shebang here - might help github to detect the used language as well

nicksinger commented on openqa-investigate:
```diff
@@ -1,391 +1,178 @@
-}
-
-caller 0 > /dev/null || main "$@"
+import requests
```
You're missing a shebang here - might help github to detect the used language as well

okurz replied:
-> https://github.com/os-autoinst/scripts/pull/332?notification_referrer_id=NT_kwDOABnW-LMxMTYzMjQ3OTAzMjoxNjkzNDMy&notifications_query=is%3Aunread#discussion_r1689552634

perlpunk commented on openqa-investigate:
```diff
@@ -1,391 +1,178 @@
+        warn(f"Command {args} failed with error: {e.stderr}")
+        return None
+
+def runjq(input_data, jq_filter):
```
Functions like this should be put into a library, not in the script directly. For bash we already have the `_common` file. We don't want to reimplement this in every script.
And I still think scripts should go into a bin folder.
The top directory is a mess.
Also tests are failing now.

okurz replied:
Sorry, please ignore openqa-investigate. That shouldn't have been part of this PR. a draft of openqa-investigate based on AI support is in
* #334

perlpunk commented on openqa-investigate:
```diff
@@ -1,391 +1,178 @@
+        result = subprocess.run(['jq', jq_filter], input=input_data, capture_output=True, text=True, check=True)
+        return result.stdout
+    except subprocess.CalledProcessError as e:
+        warn(f"jq error: {e.stderr}")
```
Will we actually get the original input here? That's what the bash function is doing currently. If jq fails we want to know what the input was.

okurz replied:
-> https://github.com/os-autoinst/scripts/pull/332?notification_referrer_id=NT_kwDOABnW-LMxMTYzMjQ3OTAzMjoxNjkzNDMy&notifications_query=is%3Aunread#discussion_r1689552634

okurz commented:
Updated:
* Simple and advanced example
* Converted single string in `subprocess.run` with `shell=True` into list (or actually a set). For that I extracted the `post` function so that I can just print in case of "dry-run" instead of the "client prefix" which was previously used to "echo" instead in bash.

Tested manually with

```
while_inotifywait sh -c "python3.11 ./openqa-advanced-retrigger-jobs -vvvv --host openqa.suse.de --failed-since \"2024-07-23T13:00\" --result failed --additional-filters \"test not like '%:investigate:%' and id=14995768\" --comment \"my foo comment\" --dry-run && python3.11 ./openqa-advanced-retrigger-jobs -vvvv --host openqa.suse.de --failed-since \"2024-07-23T13:00\" --result failed --additional-filters \"test not like '%:investigate:%' and id=14995768\" --comment \"my foo comment\""
```

Martchus commented on openqa-advanced-retrigger-jobs:
```diff
@@ -1,16 +1,105 @@
+    worker_string = f"assigned_worker_id in (select id from workers where (host='{args.worker}'{instance_string})) and " if args.worker else ""
+    additional_filters = f" and {args.additional_filters}" if args.additional_filters else ""
+
+    query = (
+        f"select id from jobs where ({worker_string}result='{args.result}' "
+        f"and clone_id is null and t_finished >= '{args.failed_since}'{additional_filters});"
+    )
```
Normally I'd say this needs escaping but of course the previous script also didn't have that…

okurz replied:
I don't know what additional escaping you mean. Why do you mean we *need* escaping?

nicksinger replied:
I think he means something in the lines of https://www.psycopg.org/psycopg3/docs/basic/params.html#execute-arguments

Currently e.g. `args.result` could be used to insert arbitrary SQL statements

nicksinger replied:
While malice is very unlikely (given you'd need access anyway) it could help if people accidentally pass arguments which cause the query to break

Martchus replied:
And one can also (accidentally) break out of the whole `psql` invocation.

Martchus commented on openqa-advanced-retrigger-jobs:
```diff
@@ -1,16 +1,105 @@
+    cmd=('openqa-cli', 'api', '--host', host, '-X', 'POST', route, *args)
+    if dry_run:
+        return print(f"dry run: {cmd}")
+    subprocess.run(cmd)
```
To retain the behavior on errors of the old script:
```suggestion
    subprocess.run(cmd, check=True)
```

Martchus commented on openqa-advanced-retrigger-jobs:
```diff
@@ -1,16 +1,105 @@
+        help="If set, only print the actions without executing",
+    )
+    args = parser.parse_args()
+    verbose_to_log = {
+        0: logging.CRITICAL,
+        1: logging.ERROR,
+        2: logging.WARN,
+        3: logging.INFO,
+        4: logging.DEBUG,
+    }
+    logging_level = logging.DEBUG if args.verbose > 4 else verbose_to_log[args.verbose]
```
Maybe this is not very explicit but I like it (considering numerical values for log levels are [documented explicitly](https://docs.python.org/3/library/logging.html#logging-levels)):
```suggestion
    logging_level = (5 - min(args.verbose, 4)) * 10
```
